### PR TITLE
Parse code by source order.

### PR DESCRIPTION
**Pull request by reitermarkus**
Extracted from https://github.com/rust-lang/rust-bindgen/pull/2369.

In order to somewhat support `#undef`, code needs to be parsed in the correct order. Previously, all macros were parsed before all variables, making it impossible to reason about whether a variable was declared before or after a macro with the same name.

### CODE REVIEW

File: bindgen/clang.rs [hunk 1]
```diff
@@ -1524,6 +1564,42 @@ impl SourceLocation {
     }
 }
 
+impl SourceLocation {
+    /// Compare source locations, also considering `#include` directives.
+    pub(crate) fn partial_cmp_with_context(
+        &self,
+        other: &Self,
+        ctx: &BindgenContext,
+    ) -> Option<cmp::Ordering> {
```
Comment (pvdrz) [inline bindgen/clang.rs:1573]: If I understood correctly, the only case where this returns `None` is if neither the file is included in the other. Given that this is only called inside `visit_sorted` wouldn't it be clearer to just return `cmp::Ordering::Equal` in that case?

File: bindgen/clang.rs [hunk 2]
```diff
@@ -1524,6 +1564,42 @@ impl SourceLocation {
     }
 }
 
+impl SourceLocation {
+    /// Compare source locations, also considering `#include` directives.
+    pub(crate) fn partial_cmp_with_context(
```
Comment (pvdrz) [inline bindgen/clang.rs:1569]: Could you expand the docs for this function stating how the comparison works? e.g: built-in locations always come first, two locations in the same file are compared by the offset, two locations in different files are compared by inclusion.

File: bindgen/clang.rs [hunk 3]
```diff
@@ -496,6 +498,44 @@ impl Cursor {
...
+                    let (source_file, _, _, offset) = location.location();
+
+                    if let Some(source_file) = source_file.name() {
+                        ctx.add_include(source_file, offset, included_file);
+                    }
+                }
+            }
+        }
+
+        children.sort_by(|child1, child2| {
```
Comment (pvdrz) [inline bindgen/clang.rs:527]: This might not be super important but what happens if `child1` is included in `child2` via another file?

File: bindgen/codegen/mod.rs [hunk 4]
```diff
@@ -274,6 +274,8 @@ struct CodegenResult<'a> {
     /// that name. This lets us give each overload a unique suffix.
     overload_counters: HashMap<String, u32>,
 
+    overload_counts: HashMap<ItemId, u32>,
```
Comment (pvdrz) [inline bindgen/codegen/mod.rs:277]: Could you document this field? Specially on why is this needed if `overload_counters` already exists.

File: bindgen/ir/context.rs [hunk 5]
```diff
@@ -356,6 +356,11 @@ pub(crate) struct BindgenContext {
     /// This needs to be an std::HashMap because the cexpr API requires it.
     parsed_macros: StdHashMap<Vec<u8>, cexpr::expr::EvalResult>,
 
+    /// A map with all include locations.
+    ///
+    /// This is needed so that items are created in the order they are defined in.
```
Comment (pvdrz) [inline bindgen/ir/context.rs:361]: could you either document to which filename each element of the tuple corresponds to or add a dedicated `struct` with named fields for this?

**Comment by pvdrz** [comment]
I just did a review and added a few comments, most of them are just about documentation. 

Could you also please document these changes in the changelog. Mainly the fact that the order of the bindings has changed and that the `__bindgen_ty_\d+` types might change name because of this?

Also I have a last question: The evaluation of define directives is not affected by this new ordering yet, right?

File: bindgen/clang.rs (hunk 3 above)
Comment (reitermarkus) [reply bindgen/clang.rs:527]: Not quite sure I understand your question. If `child1` is included in `child2` via `child3`, it will be sorted first relative to `child3`, and then relative to `child2`.


Example:

`header1.h`

```c
#define VAR1 1

#include "header2.h"

#define VAR2 2
```

`header2.h`

```c
#define VAR3 3

#include "header3.h"

#define VAR4 4

```

`header3.h`

```c
#define VAR5 5
```

will generate 

```rust
pub const VAR1: u32 = 1;
pub const VAR3: u32 = 3;
pub const VAR5: u32 = 5;
pub const VAR4: u32 = 4;
pub const VAR2: u32 = 2;
```

File: bindgen/codegen/mod.rs (hunk 4 above)
Comment (reitermarkus) [reply bindgen/codegen/mod.rs:277]: I think this can be removed again now. It was needed for sorting duringcode generation. Now, sorting is done while parsing, and it seems to work without this change.

**Comment by reitermarkus** [comment]
> The evaluation of define directives is not affected by this new ordering yet, right?

No, this is simply extracting this logic from https://github.com/rust-lang/rust-bindgen/pull/2369 for easier review.

File: bindgen/clang.rs (hunk 3 above)
Comment (pvdrz) [reply bindgen/clang.rs:527]: Yeah I think I got entangled by how the ordering works but I get it now, thanks!

**Comment by emilio** [comment]
@reitermarkus fyi, this broke Firefox (still working on a more reduced test-case). I'm a bit confused about this change tho. You completely ignore the `CXChildVisit_{Break,Continue,Recurse}` return value of the visitor? I'm surprised this didn't break more stuff.

**Comment by emilio** [comment]
I guess none of the callers cared about the different CXChildVisit but at least the API should be simplified to reflect the fact that the return value is ignored...

**Comment by pvdrz** [comment]
Hmmm... this is unfortunate. I'll try to review all the issues/PRs now.
